Support for indicating and getting feedback for e-mail test messages

lwcorp opened the pull request:
<!---Thanks for contributing to phpList!-->

## Description
1. Appended a test subject indicator to test messages (based on the existing translation of the word `test`)
1. Added a reply-to address to test messages with this logic:
   1. Left the current approach of using a manual reply-to address if one was supplied
   1. But if one wasn't supplied, take the currently logged-in admin's address
   1. If it's empty, take the general admin's address

## Screenshots (if appropriate):
![image](https://github.com/phpList/phplist3/assets/1773306/d244f819-5276-46f6-bac8-2e82ff7efe45)

lwcorp committed:
Update sendemaillib.php

1. Appended a test subject indicator to test messages
1. Added a reply-to address to test messages that have no manual reply-to: using the logged in admin's address or at least the general admin's

bramley commented:
I think this is a reasonable idea but it needs to be controlled by configuration settings so that it is optional. Something like this will allow the prefix to be specified instead of being fixed

    'test_email_prefix' => [
        'value'        => '',
        'description'  => s('Prefix for the subject of a test email'),
        'type'         => 'text',
        'allowempty'   => true,
        'category'     => 'campaign',
    ],
    'test_email_reply_to' => [
        'value'        => false,
        'description'  => s("Use the admin's email address as Reply-To of a test email"),
        'type'         => 'boolean',
        'allowempty'   => true,
        'category'     => 'campaign',
    ],

The global variable `$admin_auth` should be used to get the email address and login name of the current admin because a plugin can provide this functionality instead of core phplist

```
$adminEmail = $admin_auth->adminEmail($_SESSION['logindetails']['id']);
$adminName = $admin_auth->adminName($_SESSION['logindetails']['id']);
```
but I am not sure that it should be making the admin's login id public in this way. Maybe not bother with a name. 
The email address is a mandatory field when creating an admin, so there is no need to have a fall-back of `getConfig('admin_address')`

Also, I think that the test Reply-To should be used in preference to a Reply-To that has already been set for the campaign.

The variable `$testReplyFrom` is misnamed, it contains a Reply-To address.

lwcorp committed:
Update sendemaillib.php

Rephrased variable name

lwcorp committed:
Update sendemaillib.php

Switched to using $admin_auth

lwcorp commented:
> 'value'        => '',
> 'value'        => false,

What is the purpose value (which is either blank or false) there? In any case, so many settings seem a bit like an overkill to me, but would you like to add them to this commit? Because I'm not sure where, maybe config.php?

>Maybe not bother with a name.

I don't agree, as the more you use names, the less likely you are to get flagged as a spammer. Besides, it's only for internal testing, meaning it's the admin him/herself that interacts with the tester, so why hiding their name from the tester (who, let's face it, is probably the admin him/herself...)?

> The global variable `$admin_auth` should be used to get the email address and login name of the current admin because a plugin can provide this functionality instead of core phplist

No problem, I've just switched to it, please review.

> The email address is a mandatory field when creating an admin, so there is no need to have a fall-back of `getConfig('admin_address')`

I don't agree as in my case it's empty...
![image](https://github.com/phpList/phplist3/assets/1773306/da33f029-66ac-4c0f-be17-20c2cfc8e131)
![image](https://github.com/phpList/phplist3/assets/1773306/968bea1f-1d30-4bdb-8a94-895e7b3f004c)

> Also, I think that the test Reply-To should be used in preference to a Reply-To that has already been set for the campaign.

Do you mean it should override the campaign reply-to? I believe the purpose of testing should also be to play with different reply-to addresses just like it's possible to play with different From addresses.
 
> The variable `$testReplyFrom` is misnamed, it contains a Reply-To address.

No problem, I've just renamed, please review.

michield commented:
I've reviewed it and ran it locally, and I think it's nice. This is a nice starting point for some kind of approval work-flow. For the time being slightly manually, but maybe at some point with an "approve/disapprove" link. 

I'm not sure it needs to be configurable in the UI. We can add some optional settings in the config file and default to the current method. If you make "reply-to email" configurable, then what would you make it apart from the current admin? I guess it could go to a list or something, but that gets very messy. It makes sense to reply to the admin who sent the test email.

michield commented:
I've marked it ready to merge into the next release. @bramley if you disagree, please let us know

bramley commented:
> m not sure it needs to be configurable in the UI. We can add some optional settings in the config file and default to the current method. If you make "reply-to email" configurable, then what would you make it apart from the current admin?

What did you mean by this? I don't see any further change to use a config file setting. I envisaged a setting for the prefix with an empty value meaning don't add a prefix, and a boolean setting for the reply-to address with true meaning change the reply-to address to be that of the current admin.

michield commented:
> > m not sure it needs to be configurable in the UI. We can add some optional settings in the config file and default to the current method. If you make "reply-to email" configurable, then what would you make it apart from the current admin?
> 
> What did you mean by this? I don't see any further change to use a config file setting. I envisaged a setting for the prefix with an empty value meaning don't add a prefix, and a boolean setting for the reply-to address with true meaning change the reply-to address to be that of the current admin.

So, you mean you'd be able to avoid this new functionality by setting everything to "off"? I actually think it's quite useful to make this standard functionality, that when you check the "this is a test campaign" it adds the "(test)" to the subject. This can also be translated. 

Then we can later expand on that with some approval workflow.

petersphilo commented:
Thank you all for your continued work on this wonderful product!
it would be nice, in the future, to be able disable this new feature (both the 'test' prepend and the email address switch) via either UI or a config setting..
Huge thanks for the ability to handle emojis in the subject!!

lwcorp commented:
> Thank you all for your continued work on this wonderful product!

Thanks!

> it would be nice, in the future, to be able disable this new feature (both the 'test' prepend and the email address switch) via either UI or a config setting.

Since the official version was out days ago (after the RC was open for debate for weeks) plus it was already discussed here without a clear conclusion, may I suggest you [open a new issue](https://github.com/phpList/phplist3/issues/new) about it? You can mention #1031 (this PR) there.

DDougster commented:
Just getting ready to send our first live and my admin had a heart-attack when test showed up in the subject. I assume this is only happening when you are sending the test message prior to it going out. Was there any warning put near the test buttons reminding people of this change?

Thanks,
Doug

lwcorp commented:
It's called test, of course it's only for tests.
If you want it in the blue line, you can [submit an issue](https://github.com/phpList/phplist3/issues/new) and mention #1031 (this PR):
![image](https://github.com/phpList/phplist3/assets/1773306/a612a70b-4e4d-4370-a57e-6d78997c155e)

lofidev commented:
Firstly, thanks for all the hard work on this project.

> allow the prefix to be specified instead of being fixed

The "test" prefix is a great feature, in our case we prefer another term instead ("draft"). It would be beneficial to add this option to the config settings as @bramley [suggested](https://github.com/phpList/phplist3/pull/1031#issuecomment-1985251225).

If this is the wrong place for this, happy to open a new issue. 

Thanks,
Jay

lwcorp commented:
> Firstly, thanks for all the hard work on this project.
> Thank you all for your continued work on this wonderful product!

Thanks!!!

> > allow the prefix to be specified instead of being fixed

> If this is the wrong place for this, happy to open a new issue.

This PR was merged more than a year a go after the RC was open for debate for weeks. So please do [open a new issue](https://github.com/phpList/phplist3/issues/new) about it. You can mention #1031 (this PR) there.

But I'd like to state I've used the official language-dependent term, which was added at least 4 years ago.
What you ask will require:
1. A whole new UI form element
1. A whole new table entry
1. A whole new language-dependent term for the form element (e.g. "E-mail subject prefix for tests"), meaning dozens of translators will need to add it (although hopefully lack of support will default it to English)

So it should first be debated if the admins agree to all of this (such a debate can go in a new issue).
<img width="824" height="407" alt="image" src="https://github.com/user-attachments/assets/693e0485-1d11-456d-beba-d1cd6386c051" />
